src/passthrough/sync_io_macos.rs

```diff
@@ -0,0 +1,153 @@
+            }
+
+            // Safe because we trust the value returned by kernel.
+            unsafe { buf.set_len(res as usize) };
+
+            // Explicitly drop the lock so that it's not held while we fill in the fuse buffer.
+            mem::drop(guard);
+        }
+
+        let mut rem = &buf[..];
+        let orig_rem_len = rem.len();
+
+        while !rem.is_empty() {
+            debug_assert!(
+                rem.len() >= mem::size_of::<libc::dirent>(),
+                "fuse: not enough space left in `rem`"
+            );
+
+            let (front, back) = rem.split_at(mem::size_of::<libc::dirent>());
+
+            let dirent = unsafe { *(front.as_ptr() as *const libc::dirent) };
+
+            let namelen = dirent.d_namlen as usize;
+            debug_assert!(
+                namelen <= back.len(),
+                "fuse: back is smaller than `namelen`"
+            );
+
+            let name = &back[..namelen];
+            let res = if name.starts_with(CURRENT_DIR_CSTR) || name.starts_with(PARENT_DIR_CSTR) {
+                Ok(1)
+            } else {
+                let name = bytes_to_cstr(name)
+                    .map_err(|e| {
+                        error!("fuse: do_readdir: {:?}", e);
+                        einval()
+                    })?
+                    .to_bytes();
+
+                add_entry(
+                    DirEntry {
+                        ino: dirent.d_ino,
+                        #[cfg(target_os = "linux")]
+                        offset: dirent.d_seekoff as u64,
+                        #[cfg(target_os = "macos")]
+                        offset: dirent.d_seekoff,
+                        type_: dirent.d_type as u32,
+                        name,
+                    },
+                    data.borrow_fd().as_raw_fd(),
+                )
+            };
+
+            debug_assert!(
+                rem.len() >= dirent.d_reclen as usize,
+                "fuse: rem is smaller than `d_reclen`"
+            );
+
+            match res {
+                Ok(0) => break,
+                Ok(_) => rem = &rem[dirent.d_reclen as usize..],
+                Err(e) if rem.len() == orig_rem_len => return Err(e),
+                Err(_) => return Ok(()),
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn do_getattr(
+        &self,
+        inode: Inode,
+        handle: Option<Handle>,
+    ) -> io::Result<(LibCStat, Duration)> {
+        let st;
+        let data = self.inode_map.get(inode).map_err(|e| {
+            error!("fuse: do_getattr ino {} Not find err {:?}", inode, e);
+            e
+        })?;
+
+        // kernel sends 0 as handle in case of no_open, and it depends on fuse server to handle
+        // this case correctly.
+        if !self.no_open.load(Ordering::Relaxed) && handle.is_some() {
+            // Safe as we just checked handle
+            let hd = self.handle_map.get(handle.unwrap(), inode)?;
+            st = stat(hd.get_file());
```

Comment:
The macos `do_getattr` differs from linux `do_getattr` only at this line `st = stat_fd(hd.get_file(), None);` ?

Then maybe it's better to wrap `st = stat_fd(hd.get_file(), None);` and `st = stat(hd.get_file());` to platform specific function such as `stat_fd`

Reply:
done